### DISCUSSION TITLE
is it possible that mutation action in loader function

### DISCUSSION THREAD

**Discussion by WreewanMorhee** [remix-run/remix#7753] [Q&A] [answered]
like title 
is it possible ??

**Comment by sergiodxa** [comment]
I'm not sure I understand your question.

Do you want to run a mutation in a loader? You can do anything there although it's intended for reading data.

Or do you want to change the code in the action from within the loader?

**Reply by WreewanMorhee** [reply]
I can use HTML Form or useFetcher to call action in component
Is there a way that I could call action in loader ?

**Reply by sergiodxa** [reply, marked as answer]
The simplest way would be to move the code in the action to a different function and call it in both places
```ts
async function doSomething() { /* code */ }

export async function loader(args: LoaderFunctionArgs) {
  // some code
  await doSomething()
  // more code
}

export async function action(args: ActionFunctionArgs) {
  // some code
  await doSomething()
  // more code
}
```

This way, your `doSomething` function can receive exactly the arguments it needs and return values instead of responses which gives you more flexibility when using it.

Another option is to call the action

```ts
export async function action(args: ActionFunctionArgs) { /* some code */ }

export async function loader({ request, params, context }: LoaderFunctionArgs) {
  let body = new FormData()
  // add values to body here
  let actionResponse = await action({
    request: new Request(request.url, { method: "post", headers: request.headers, body }),
    params,
    context,
  })
  let actionData = await actionResponse.json()
  // more code using actionData here
}
```
And finally another option is to do a fetch to the action
```ts
export async function action(args: ActionFunctionArgs) { /* some code */ }

export async function loader({ request, params, context }: LoaderFunctionArgs) {
  let actionResponse = await fetch(request.url, {
    method: "POST",
    headers
  })

  // more code using actionResponse here
}
```
In all cases where I used `request.url` that's supposing the action is on the same URL, but if it's in a different URL you can do
```ts
let url = new URL("/some/pathname", request.url)
```
And this way change the pathname while retaining the hostname and protocol.

**Reply by WreewanMorhee** [reply]
is it possible that:
A loader call B action ??
example below:

route.a.tsx
```
export async function loader({ request, params, context }: LoaderFunctionArgs) {
  // some thing
}
```


route.b.tsx
```
export async function action(args: ActionFunctionArgs) { /* some code */ }
```

loader in route.a.tsx want to call the action in route.b.tsx
is it possible ?

**Reply by sergiodxa** [reply]
I would do the first thing and move the logic I want to execute in both places to a function I can re-use, but you can also import the route B action in the route A file, or you could fetch it.

**Reply by WreewanMorhee** [reply]
thank u sir 
I think I know how to do it
